File: Try_GameEngine5/src/com/example/try_gameengine/action/MovementActionItemCountDownTimer.java
package com.example.try_gameengine.action;

import java.util.ArrayList;
import java.util.List;

import com.example.try_gameengine.action.visitor.IMovementActionVisitor;
import android.os.CountDownTimer;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

/**
 * MovementActionItem is a item(leaf) in MovementAction composites.
 * 
 * @author irons
 * 
 */
public class MovementActionItemCountDownTimer extends MovementActionItemForMilliseconds {
	CountDownTimer countDownTimer;
	
	/**
	 * constructor.
	 * 
	 * @param millisTotal
	 * 			milliseconds for whole action running.
	 * @param millisDelay
	 * 			milliseconds for delay.
	 * @param dx
	 * 			x-dir move for per delay time.
	 * @param dy
	 * 			y-dir move for per delay time.
	 */
	public MovementActionItemCountDownTimer(long millisTotal, long millisDelay, final int dx,
			final int dy) {
		this(millisTotal, millisDelay, dx, dy, "MovementItem");
	}

	/**
	 * constructor.
	 * 
	 * @param millisTotal
	 * 			milliseconds for whole action running.
	 * @param millisDelay
	 * 			milliseconds for delay.
	 * @param dx
	 * 			x-dir move for per delay time.
	 * @param dy
	 * 			y-dir move for per delay time.
	 * @param description
	 * 			description for this movement action.
	 */
	public MovementActionItemCountDownTimer(long millisTotal, long millisDelay, final int dx,
			final int dy, String description) {
		super(new MovementActionInfo(millisTotal, millisDelay, dx, dy, description));
	}

	/**
	 * constructor.
	 * 
	 * @param info
	 */
	public MovementActionItemCountDownTimer(MovementActionInfo info) {
		super(info);
	}

	@Override
	public void start() {
		// TODO Auto-generated method stub
		if (isActionFinish) {
			isReset = true;
			thread = new Thread(new Runnable() {

				@Override
				public void run() {
					// TODO Auto-generated method stub
					while (isReset) {
						isReset = false;
						synchronized (MovementActionItemCountDownTimer.this) {
							countDownTimer.start();
							try {
								MovementActionItemCountDownTimer.this.wait();
							} catch (InterruptedException e) {
								// TODO Auto-generated catch block
								e.printStackTrace();
								isActionFinish = false;
							}
						}

					}
				}
			});
			thread.start();
		}

		isActionFinish = false;
		if (isFirstTime) {
			resetTotal = millisTotal;
			isFirstTime = false;

			thread = new Thread(new Runnable() {

				@Override
				public void run() {
					// TODO Auto-generated method stub
					while (isReset) {
						isReset = false;
						synchronized (MovementActionItemCountDownTimer.this) {
							countDownTimer.start();
							try {
								MovementActionItemCountDownTimer.this.wait();
							} catch (InterruptedException e) {
								// TODO Auto-generated catch block
								e.printStackTrace();
								isActionFinish = false;
							}
						}

					}
				}
			});
			thread.start();
		}

		if (info.getSprite() != null)
			info.getSprite().setAction(info.getSpriteActionName());

	}

	@Override
	protected MovementAction initTimer(){ super.initTimer();
		millisTotal = info.getTotal();
		millisDelay = info.getDelay();
		dx = info.getDx();
		dy = info.getDy();
		rotationController = info.getRotationController();
		gravityController = info.getGravityController();

		countDownTimer = new CountDownTimer(millisTotal, millisDelay) {

			@Override
			public void onTick(long millisUntilFinished) {
				Log.e("t", millisUntilFinished + "");
				Log.e("t", millisUntilFinished / 1000 + "");

				doRotation();
				doGravity();
				Log.e("dx", dx + "");
				Log.e("dy", dy + "");

				resumeTotal = millisUntilFinished;
				timerOnTickListener.onTick(dx, dy);
			}

			@Override
			public void onFinish() {
				if (isLoop) {
					handler.sendEmptyMessage(0);
					Log.e("Timer", "loop");
				} else {
					synchronized (MovementActionItemCountDownTimer.this) {
						MovementActionItemCountDownTimer.this.notifyAll();
					}
					doReset();
					isActionFinish = true;
					Log.e("Timer", "finish");
				}
			}
		};
		return this;
	}

	/**
	 * If rotationController is not null do rotation execute.
	 */
	private void doRotation() {
		if (rotationController != null) {
			rotationController.execute(info);
			dx = info.getDx();
			dy = info.getDy();
		}
	}

	/**
	 * If gravityController is not null do gravity execute.
	 */
	private void doGravity() {
		if (gravityController != null) {
			gravityController.execute(info);
			dx = info.getDx();
			dy = info.getDy();
		}
	}

	/**
	 * reset action.
	 */
	private void doReset() {
		if (gravityController != null) {
			gravityController.reset(info);
		}
		if (rotationController != null)
			rotationController.reset(info);

		millisTotal = info.getTotal();
		millisDelay = info.getDelay();
		dx = info.getDx();
		dy = info.getDy();
		initTimer();
	}

	@Override
	public MovementAction getAction() {
		return this;
	}

	@Override
	public List<MovementAction> getActions() {
		return actions;
	}

	@Override
	public MovementActionInfo getInfo() {
		return info;
	}

	@Override
	public void setInfo(MovementActionInfo info) {
		this.info = info;
	}

	@Override
	public List<MovementAction> getCurrentActionList() {
		List<MovementAction> actions = new ArrayList<MovementAction>();
		actions.add(this);
		return actions;
	}

	@Override
	public List<MovementActionInfo> getCurrentInfoList() {
		List<MovementActionInfo> infos = new ArrayList<MovementActionInfo>();
		infos.add(this.info);
		return infos;
	}

	@Override
	public List<MovementActionInfo> getMovementInfoList() {
		List<MovementActionInfo> infos = new ArrayList<MovementActionInfo>();
		infos.add(this.info);
		return infos;
	}

	@Override
	public void cancelMove() {
		countDownTimer.cancel();
	}

	/**
	 * handler is use for old movement action witch use timer.
	 */
	Handler handler = new Handler(Looper.getMainLooper()) {
		public void handleMessage(android.os.Message msg) {
			initTimer();
			info.setTotal(resetTotal);
			isReset = true;
			thread.interrupt();
			start();
		};
	};

	@Override
	void pause() {
		new Thread(new Runnable() {

			@Override
			public void run() {
				if (!isActionFinish) {
					countDownTimer.cancel();

					try {
						Thread.sleep(800);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}

					millisTotal = resumeTotal;
					info.setTotal(millisTotal);
					handler.sendEmptyMessage(0);
				}
			}
		}).start();
	}

	@Override
	public boolean isFinish() {
		return isActionFinish;
	}

	@Override
	public void accept(IMovementActionVisitor movementActionVisitor) {
		movementActionVisitor.visitLeaf(this);
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/action/MovementActionItemCountDownTimer.kt
package com.example.try_gameengine.action

import android.os.CountDownTimer
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import com.example.try_gameengine.action.visitor.IMovementActionVisitor

/**
 * MovementActionItem is a item(leaf) in MovementAction composites.
 * 
 * @author irons
 // */
class MovementActionItemCountDownTimer : MovementActionItemForMilliseconds {
    var countDownTimer: CountDownTimer? = null

    /**
     * constructor.
     * 
     * @param millisTotal
     * milliseconds for whole action running.
     * @param millisDelay
     * milliseconds for delay.
     * @param dx
     * x-dir move for per delay time.
     * @param dy
     * y-dir move for per delay time.
     * @param description
     * description for this movement action.
     // */
    /**
     * constructor.
     * 
     * @param millisTotal
     * milliseconds for whole action running.
     * @param millisDelay
     * milliseconds for delay.
     * @param dx
     * x-dir move for per delay time.
     * @param dy
     * y-dir move for per delay time.
     // */
    @JvmOverloads
    constructor(
        millisTotal: Long, millisDelay: Long, dx: Int,
        dy: Int, description: String? = "MovementItem"
    ) : super(MovementActionInfo(millisTotal, millisDelay, dx.toFloat(), dy.toFloat(), description))

    /**
     * constructor.
     * 
     * @param info
     // */
    constructor(info: MovementActionInfo?) : super(info ?: MovementActionInfo(0, 0, 0f, 0f))

    public override fun start() {
        // TODO Auto-generated method stub
        if (isActionFinish) {
            isReset = true
            thread = Thread(object : Runnable {
                override fun run() {
                    // TODO Auto-generated method stub
                    while (isReset) {
                        isReset = false
                        synchronized(this@MovementActionItemCountDownTimer) {
                            countDownTimer!!.start()
                            try {
                                (this@MovementActionItemCountDownTimer as Object).wait()
                            } catch (e: InterruptedException) {
                                // TODO Auto-generated catch block
                                e.printStackTrace()
                                isActionFinish = false
                            }
                        }
                    }
                }
            })
            thread!!.start()
        }

        isActionFinish = false
        if (isFirstTime) {
            resetTotal = millisTotal
            isFirstTime = false

            thread = Thread(object : Runnable {
                override fun run() {
                    // TODO Auto-generated method stub
                    while (isReset) {
                        isReset = false
                        synchronized(this@MovementActionItemCountDownTimer) {
                            countDownTimer!!.start()
                            try {
                                (this@MovementActionItemCountDownTimer as Object).wait()
                            } catch (e: InterruptedException) {
                                // TODO Auto-generated catch block
                                e.printStackTrace()
                                isActionFinish = false
                            }
                        }
                    }
                }
            })
            thread!!.start()
        }

        if (info.getSprite() != null) info.getSprite().setAction(info.getSpriteActionName())
    }

    public override fun initTimer(): MovementAction {
        super.initTimer()
        millisTotal = info.getTotal()
        millisDelay = info.getDelay()
        dx = info.getDx()
        dy = info.getDy()

        initCountDownTimer()
        return this
    }

    private fun initCountDownTimer() {
        countDownTimer = object : CountDownTimer(millisTotal, millisDelay) {
            override fun onTick(millisUntilFinished: Long) {
                Log.e("t", millisUntilFinished.toString() + "")
                Log.e("t", (millisUntilFinished / 1000).toString() + "")
                Log.e("dx", dx.toString() + "")
                Log.e("dy", dy.toString() + "")

                resumeTotal = millisUntilFinished
                timerOnTickListener!!.onTick(dx, dy)
            }

            override fun onFinish() {
                if (isLoop) {
                    handler.sendEmptyMessage(0)
                    Log.e("Timer", "loop")
                } else {
                    synchronized(this@MovementActionItemCountDownTimer) {
                        (this@MovementActionItemCountDownTimer as Object).notifyAll()
                    }
                    doReset()
                    isActionFinish = true
                    Log.e("Timer", "finish")
                }
            }
        }
    }

    /**
     * reset action.
     // */
    private fun doReset() {
        millisTotal = info.getTotal()
        millisDelay = info.getDelay()
        dx = info.getDx()
        dy = info.getDy()

//		initTimer();
        initCountDownTimer()
    }

    override fun trigger() {
        // TODO Auto-generated method stub
    }

    public override fun getAction(): MovementAction {
        return this
    }

    public override fun getActions(): MutableList<MovementAction> {
        return actions
    }

    public override fun getInfo(): MovementActionInfo {
        return info
    }

    public override fun setInfo(info: MovementActionInfo?) {
        this.info = info ?: return
    }

    public override fun getCurrentActionList(): MutableList<MovementAction> {
        val actions: MutableList<MovementAction> = ArrayList<MovementAction>()
        actions.add(this)
        return actions
    }

    public override fun getCurrentInfoList(): MutableList<MovementActionInfo?> {
        val infos: MutableList<MovementActionInfo?> = ArrayList<MovementActionInfo?>()
        infos.add(this.info)
        return infos
    }

    public override fun getMovementInfoList(): MutableList<MovementActionInfo?> {
        val infos: MutableList<MovementActionInfo?> = ArrayList<MovementActionInfo?>()
        infos.add(this.info)
        return infos
    }

    public override fun cancelMove() {
        countDownTimer!!.cancel()
    }

    /**
     * handler is use for old movement action witch use timer.
     // */
    var handler: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            initTimer()
            info.setTotal(resetTotal)
            isReset = true
            thread!!.interrupt()
            start()
        }
    }

    override fun pause() {
        Thread(object : Runnable {
            override fun run() {
                if (!isActionFinish) {
                    countDownTimer!!.cancel()

                    try {
                        Thread.sleep(800)
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }

                    millisTotal = resumeTotal
                    info.setTotal(millisTotal)
                    handler.sendEmptyMessage(0)
                }
            }
        }).start()
    }

    public override fun isFinish(): Boolean {
        return isActionFinish
    }

    public override fun accept(movementActionVisitor: IMovementActionVisitor) {
        movementActionVisitor.visitLeaf(this)
    }
}
